src/library/scala/collection/immutable/List.scala

```diff
@@ -259,29 +259,26 @@ sealed abstract class List[+A]
       h
     }
   }
+
   final override def flatMap[B](f: A => IterableOnce[B]): List[B] = {
-    if (this eq Nil) Nil else {
-      var rest = this
-      var found = false
-      var h: ::[B] = null
-      var t: ::[B] = null
-      while (rest ne Nil) {
-        f(rest.head).iterator.foreach { b =>
-          if (!found) {
-            h = new ::(b, Nil)
-            t = h
-            found = true
-          }
-          else {
-            val nx = new ::(b, Nil)
-            t.next = nx
-            t = nx
-          }
+    var rest = this
+    var h: ::[B] = null
+    var t: ::[B] = null
+    while (rest ne Nil) {
+      f(rest.head).iterator.foreach { b =>
```

Comment:
It's usually faster to use `next/hasNext` to step through the iterator than to use `foreach`, especially since the closure here has to change `h` and `t` to Var classes.

Reply:
ok, I'll adjust